### PR TITLE
Cb save image link

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [ ] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by kimiamajd98**
Closes #2 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

**Comment by sarahsehr** [comment]
> The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

It looks like some of these items still need to be completed.

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   def home; end
+  def form; end
```
Comment (sarahsehr) [inline app/controllers/application_controller.rb:5]: Will it be problematic to continue to add new actions to this controller? What would be a good way to organize the various actions?

Is `form` a descriptive enough name? Will all forms in this application use this same action? If not, how will we tell this one from others?

Comment (kimiamajd98) [reply app/controllers/application_controller.rb:5]: Oops! I should have deleted form. I stopped using form once I turned image into an action and I put the rest of the methods into the image controller. Will remove form and commit right now!

File: app/views/application/form.html.erb [hunk 2]
```diff
@@ -1,2 +1,11 @@
 <h1>Image Form</h1>
 
+<%= form_with model: @image do |form| %>
+  <div>
+    <%= form.label :url %><br>
```
Comment (sarahsehr) [inline app/views/application/form.html.erb]: When should you use `br` vs `div`?

File: config/routes.rb [hunk 3]
```diff
@@ -2,5 +2,6 @@
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root 'application#home'
   get '/form', to: 'application#form', as: 'form'
+  resources :images
```
Comment (sarahsehr) [inline config/routes.rb:4]: Do you need all the routes that this will generate?

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+class ImagesController < ActionController::Base
+  def new
+    @image = Image.new
+  end
+    
+  def create
```
Comment (sarahsehr) [inline app/controllers/images_controller.rb:10]: This controller is committed without a test. Do you see any consequences to this?

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    @image = Image.new(url: '...')
+    if @image.save
+      redirect_to @image
+    else
+      render :new
+    end
+  end
+end
+
```
Comment (sarahsehr) [inline app/controllers/images_controller.rb]: This little circle with a line through it has meaning, do you know what it is?

File: app/models/image.rb [hunk 6]
```diff
@@ -1,2 +1,3 @@
 class Image < ApplicationRecord
+  validates :url, presence: true, url : true
```
Comment (sarahsehr) [inline app/models/image.rb]: Test?

Comment (sarahsehr) [reply app/models/image.rb]: Is this valid Ruby?

File: app/controllers/images_controller.rb [hunk 7]
```diff
@@ -1,10 +1,14 @@
 class ImagesController < ActionController::Base
+
+  def show
+    @image = Image.find(params[:id])
```
Comment (sarahsehr) [inline app/controllers/images_controller.rb:3]: Test?

File: app/controllers/images_controller.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+class ImagesController < ActionController::Base
```
Comment (sarahsehr) [inline app/controllers/images_controller.rb:1]: Is this the parent class that the tutorials say to use?

File: app/views/application/home.html.erb [hunk 9]
```diff
@@ -1,2 +1,2 @@
 <h1>Hello World</h1>
-<%=link_to "Image Submission Form", form_path%>
+<%=link_to "Image Submission Form", new_image_path%>
```
Comment (sarahsehr) [inline app/views/application/home.html.erb:2]: Does this line follow Ruby styles?

File: config/routes.rb [hunk 10]
```diff
@@ -1,7 +1,6 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root 'application#home'
-  get '/form', to: 'application#form', as: 'form'
```
Comment (sarahsehr) [inline config/routes.rb]: Do we still need the form action in ApplicationController?

File: app/views/images/show.html.erb [hunk 11]
```diff
@@ -0,0 +1,2 @@
+<%= image_tag(@image.url) %>
+<h1><%= @image.url %><\h1>
```
Comment (sarahsehr) [inline app/views/images/show.html.erb]: What is the semantic meaning of an `h1`?

File: config/routes.rb (hunk 10 above)
Comment (kimiamajd98) [reply config/routes.rb]: No, it's not in the latest commit!

File: app/models/image.rb [hunk 12]
```diff
@@ -1,3 +1,6 @@
 class Image < ApplicationRecord
-  validates :url, presence: true
+  validates :url, presence: true, format: {
```
Comment (sarahsehr) [inline app/models/image.rb:2]: How can we verify that this works the way we expect?

File: app/controllers/application_controller.rb [hunk 13]
```diff
@@ -1,5 +1,6 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   def home; end
+
```
Comment (sarahsehr) [inline app/controllers/application_controller.rb]: Are there any consequences to these style changes being in their own commit?

File: app/views/images/show.html.erb (hunk 11 above)
Comment (kimiamajd98) [reply app/views/images/show.html.erb]: It's a primary header. That line was just for my own testing purposes. I was having trouble getting the image to show and wanted to see if I could even get the url to show and I accidentally committed it. That line isn't in the latest commit!

**Review by sarahsehr — CHANGES_REQUESTED** [review]
I've added a few questions to think about. Most are trying to hint at a change that should be made.

One other question. What is the role of the `ApplicationController` class? Why does it exist and how and when would it be used?

File: app/views/application/form.html.erb (hunk 2 above)
Comment (kimiamajd98) [reply app/views/application/form.html.erb]: br should be used as a line break while div is used to signify a block level element. This should probably be wrapped in a div rather than a br.

File: config/routes.rb (hunk 3 above)
Comment (kimiamajd98) [reply config/routes.rb:4]: No I only need index, show, new, destroy, and create

File: app/controllers/images_controller.rb (hunk 4 above)
Comment (kimiamajd98) [reply app/controllers/images_controller.rb:10]: added tests!

File: app/controllers/images_controller.rb (hunk 5 above)
Comment (kimiamajd98) [reply app/controllers/images_controller.rb]: newline needed at end of file, fixed!

File: app/models/image.rb (hunk 6 above)
Comment (kimiamajd98) [reply app/models/image.rb]: Added image model tests and this validation was wrong, I updated it in a following commit

File: app/controllers/images_controller.rb (hunk 7 above)
Comment (kimiamajd98) [reply app/controllers/images_controller.rb:3]: added tests

File: app/models/image.rb (hunk 12 above)
Comment (kimiamajd98) [reply app/models/image.rb:2]: I added tests for the image model that verify that only valid / nonblank url's will be allowed

File: app/controllers/images_controller.rb (hunk 8 above)
Comment (kimiamajd98) [reply app/controllers/images_controller.rb:1]: No, the parent class should be application controller. Fixed!

File: app/controllers/application_controller.rb (hunk 13 above)
Comment (kimiamajd98) [reply app/controllers/application_controller.rb]: The style issues caused some of the circleCI tests to fail. I should have run rubocop before committing so the style changes weren't in a separate commit.

File: app/views/application/home.html.erb (hunk 9 above)
Comment (kimiamajd98) [reply app/views/application/home.html.erb:2]: This one I'm not sure of.. I thought new_image_path was correct. Should it be new_image_url instead?

File: app/controllers/application_controller.rb (hunk 13 above)
Comment (sarahsehr) [reply app/controllers/application_controller.rb]: Besides having to do all the work at the end, this can also affect future dev productivity. If someone is using Annotate (git "blame") to see what the history of a particular line, what they will see in style changes and not the actual changes in behavior that they are looking for. By running rubocop before committing each commit OR fixing up the styles changes into the correct commit, we can avoid this second problem.

File: app/controllers/images_controller.rb [hunk 14]
```diff
@@ -12,7 +12,7 @@ def create
     if @image.save
       redirect_to @image
     else
-      render :new
+      render :new, status: :unprocessable_entity
```
Comment (sarahsehr) [inline app/controllers/images_controller.rb:15]: Good

File: app/views/images/new.html.erb [hunk 15]
```diff
@@ -1,4 +1,4 @@
-<h1>Image Form</h1>
+<h1 id = 'header'>Image Form</h1>
```
Comment (sarahsehr) [inline app/views/images/new.html.erb:1]: Does this work correctly with the spaces? I'm used to seeing this like `<h1 id='header'>Image Form</h1>`

File: test/controllers/images_controller_test.rb [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+    assert_response :ok
+    assert_select 'img.img[src=?]', @image.url
+  end
+
+  def test_new
+    get new_image_path
+
+    assert_response :ok
+    assert_select '#header', 'Image Form'
```
Comment (sarahsehr) [inline test/controllers/images_controller_test.rb:19]: Is there anything else we can assert?

File: test/controllers/images_controller_test.rb [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  def test_create__succeed
+    assert_difference('Image.count', 1) do
```
Comment (sarahsehr) [inline test/controllers/images_controller_test.rb:23]: Ruby style most commonly drops the parenthesis. `assert_difference 'Image.count', 1 do`

File: test/controllers/images_controller_test.rb [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+      image_params = { url: 'https://some-url.png' }
+      post images_path, params: { image: image_params }
+    end
```
Comment (sarahsehr) [inline test/controllers/images_controller_test.rb:26]: Line 23 asserts that a new Image is created. Is there anything else that might be helpful to check about that creation?

File: app/views/images/new.html.erb (hunk 15 above)
Comment (kimiamajd98) [reply app/views/images/new.html.erb:1]: I think it did work correctly because the test passed but I will remove the spaces if that is the more common practice!

File: test/controllers/images_controller_test.rb (hunk 18 above)
Comment (kimiamajd98) [reply test/controllers/images_controller_test.rb:26]: I added an assertion to check if the url of the image just created matches the intended url. Not sure if there is anything else I should assert?

File: test/models/image_test.rb [hunk 19]
```diff
@@ -1,7 +1,23 @@
 require 'test_helper'
 
 class ImageTest < ActiveSupport::TestCase
-  # test "the truth" do
-  #   assert true
-  # end
+  def test_image__valid
```
Comment (sarahsehr) [inline test/models/image_test.rb:4]: Test coverage on the unit level, like this model test, should be very thorough. That is, I shouldn't be able to delete any code in the model without a test failing. Do you have sufficient test coverage that I could do that?

File: test/controllers/images_controller_test.rb [hunk 20]
```diff
@@ -17,19 +17,23 @@ def test_new
 
     assert_response :ok
     assert_select '#header', 'Image Form'
+    assert_select 'form'
+    assert_select 'form input[name=?]', 'image[url]'
```
Comment (sarahsehr) [inline test/controllers/images_controller_test.rb:21]: You don't need to change this, but for future reference. `assert_select` supports nested assertions:

```
assert_select 'form' do |form|
  assert_select form, 'input[name=?]', 'image[url]'
  assert_select form, 'input[value=?]', 'iCreate Image'
end
```

I find this example reads a little bit more clearly, but what you have is not wrong.

File: app/views/application/home.html.erb (hunk 9 above)
Comment (sarahsehr) [reply app/views/application/home.html.erb:2]: `<%= link_to "Image Submission Form", new_image_path %>`

I was referring to the spacing.

File: test/controllers/images_controller_test.rb [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+class ImagesControllerTest < ActionDispatch::IntegrationTest
+  setup do
+    @image = Image.create!(url: 'https://url.jpg')
+  end
+
+  def test_show
+    get image_path(@image.id)
+
+    assert_response :ok
+    assert_select 'img.img[src=?]', @image.url
```
Comment (sarahsehr) [inline test/controllers/images_controller_test.rb:12]: One thing that I prefer is to use string literals as much as reasonably possible in tests.

Image that you had a typo, so your code was something like this:

```
setup do
    @image = Image.create!(ur: 'https://url.jpg')
  end

  def test_show
    get image_path(@image.id)

    assert_response :ok
    assert_select 'img.img[src=?]', @image.url
end
```

Let's assume there are no validations on `url` presence, so `@image` would be successfully created, but `@image.url` would actually be `nil`. So now line 12 will pass, but we don't actually have a value for `src`!

You may think that it is far-fetched that you could have a typo in the attribute name and that there would be no validations. However, you'll find that most apps use special test objects called "exemplars" that give us nice defaults for things, so it's easier have typo or misassignment of an attribute without it being detected.

TL;DR, I would change this line to 
```
assert_select 'img.img[src=?]', 'https://url.jpg'
```

File: test/controllers/images_controller_test.rb [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def test_create__succeed
+    assert_difference 'Image.count', 1 do
+      image_params = { url: 'https://some-url.png' }
+      post images_path, params: { image: image_params }
+    end
+
+    assert_redirected_to image_path(Image.last)
+    assert_equal('https://some-url.png', Image.last.url)
```
Comment (sarahsehr) [inline test/controllers/images_controller_test.rb:32]: Do you see any performance implications with calling `Image.last` twice?

File: test/controllers/images_controller_test.rb [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def test_create__fail
+    assert_no_difference 'Image.count' do
+      image_params = { url: '' }
+      post images_path, params: { image: image_params }
+    end
+
+    assert_response :unprocessable_entity
```
Comment (sarahsehr) [inline test/controllers/images_controller_test.rb:41]: How do we know that `new` is being rendered?

File: test/controllers/images_controller_test.rb (hunk 22 above)
Comment (kimiamajd98) [reply test/controllers/images_controller_test.rb:32]: It's probably better to save Image.last in a local variable so that I don't need to call last twice

File: test/controllers/images_controller_test.rb (hunk 23 above)
Comment (kimiamajd98) [reply test/controllers/images_controller_test.rb:41]: I added assertions to check that the form is being rendered

File: test/models/image_test.rb (hunk 19 above)
Comment (kimiamajd98) [reply test/models/image_test.rb]: I added tests to cover the other image extensions. It seems to me that all the code in the images file has a test where if the code was deleted, a test would fail.

**Review by sarahsehr — APPROVED** [review]
Nice work! This looks quite clean. :)

Let's not worry about it for this story, but one of the things that would be good for the next story is to make commits that will be more helpful as a history (putting related things together). This would be a great topic to discuss with Joe.
